spec/api/labwares_spec.rb

```diff
@@ -6,8 +6,218 @@
+        expect(json["errors"][0]).to eq "User is not authorised"
+      end
+
+      it "will not be valid when the location does not exist" do
+        payload = {
+          user_code: technician.barcode,
+          labwares: [
+            { location_barcode: "a invalid location", labware_barcode: labware_barcode1 }
+          ]
+        }
+        expect do
+          post api_labwares_path, params: payload
+        end.to change(Labware, :count).by(0)
+
+        expect(response).to have_http_status(:unprocessable_entity)
+        json = ActiveSupport::JSON.decode(response.body)
+        expect(json["errors"]).not_to be_empty
+        expect(json["errors"].length).to eq 1
+        expect(json["errors"][0]).to eq "location(s) with barcode 'a invalid location' do not exist"
+      end
+
+      it "will not be valid when the location does not have a parent" do
+        payload = {
+          user_code: technician.barcode,
+          labwares: [
+            { location_barcode: location_no_parent.barcode, labware_barcode: labware_barcode1 }
+          ]
+        }
+        expect do
+          post api_labwares_path, params: payload
+        end.to change(Labware, :count).by(0)
+
+        expect(response).to have_http_status(:unprocessable_entity)
+        json = ActiveSupport::JSON.decode(response.body)
+        expect(json["errors"]).not_to be_empty
+        expect(json["errors"].length).to eq 1
+        expect(json["errors"][0]).to eq "Validation failed: Location does not have a parent"
+      end
+
+      it "will not be valid when there is some data missing" do
+        payload = {
+          user_code: technician.barcode,
+          labwares: [
+            { location_barcode: '', labware_barcode: labware_barcode1 }
+          ]
+        }
+        expect do
+          post api_labwares_path, params: payload
+        end.to change(Labware, :count).by(0)
+
+        expect(response).to have_http_status(:unprocessable_entity)
+        json = ActiveSupport::JSON.decode(response.body)
+        expect(json["errors"]).not_to be_empty
+        expect(json["errors"].length).to eq 2
+        expect(json["errors"][0]).to eq "location(s) with barcode '' do not exist"
+        expect(json["errors"][1]).to eq "It looks like there is some missing or invalid data. Please review and remove anything that shouldn't be there."
+      end
+
+      it "will not be valid when no labwares are provided" do
+        payload = {
+          user_code: technician.barcode
+        }
+        expect do
+          post api_labwares_path, params: payload
+        end.to change(Labware, :count).by(0)
+
+        expect(response).to have_http_status(:unprocessable_entity)
+        json = ActiveSupport::JSON.decode(response.body)
+        expect(json["errors"]).not_to be_empty
+        expect(json["errors"].length).to eq 1
+        expect(json["errors"][0]).to eq "No labwares have been provided"
+      end
+
+      # duplication location?
+      # duplication labware?
```

Comment:
Maybe these are just notes you left for yourself?  If not, could they either be made more elaborate so someone else knows what they mean / need to do?  As they stand, I don't know what they're trying to tell me when reading them.

Reply:
Yes, resolved notes for myself!